<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Лемматизация-твитов" data-toc-modified-id="Лемматизация-твитов-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Лемматизация твитов</a></span></li><li><span><a href="#Определим-признаки-и-целевой-признак" data-toc-modified-id="Определим-признаки-и-целевой-признак-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Определим признаки и целевой признак</a></span></li><li><span><a href="#Создадим-мешок-слов-с-учётом-стоп-слов" data-toc-modified-id="Создадим-мешок-слов-с-учётом-стоп-слов-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Создадим мешок слов с учётом стоп-слов</a></span></li><li><span><a href="#Проверим-распределение-между-классами" data-toc-modified-id="Проверим-распределение-между-классами-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Проверим распределение между классами</a></span></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#Линейная-регрессия.-Подбор-гиперпараметров" data-toc-modified-id="Линейная-регрессия.-Подбор-гиперпараметров-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Линейная регрессия. Подбор гиперпараметров</a></span></li><li><span><a href="#Обучение-модели-LogisticRegression-на-подобранных-гиперпараметрах" data-toc-modified-id="Обучение-модели-LogisticRegression-на-подобранных-гиперпараметрах-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обучение модели LogisticRegression на подобранных гиперпараметрах</a></span></li><li><span><a href="#Модель-DecisionTreeClassifier.-Подбор-гипперпараметров" data-toc-modified-id="Модель-DecisionTreeClassifier.-Подбор-гипперпараметров-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Модель DecisionTreeClassifier. Подбор гипперпараметров</a></span></li><li><span><a href="#Обучение-модели-DecisionTreeClassifier-на-подобранных-гиперпараметрах" data-toc-modified-id="Обучение-модели-DecisionTreeClassifier-на-подобранных-гиперпараметрах-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Обучение модели DecisionTreeClassifier на подобранных гиперпараметрах</a></span></li><li><span><a href="#Обучение-модели-CatBoostClassifier" data-toc-modified-id="Обучение-модели-CatBoostClassifier-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обучение модели CatBoostClassifier</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span><ul class="toc-item"><li><span><a href="#Проверка-линейной-регрессии-на-тестовых-данных." data-toc-modified-id="Проверка-линейной-регрессии-на-тестовых-данных.-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Проверка линейной регрессии на тестовых данных.</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Подготовка

In [1]:
#Подключим библиотеки
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import nltk
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from pymystem3 import Mystem
from nltk.corpus import stopwords 
from sklearn.feature_extraction.text import TfidfVectorizer 
from nltk.corpus import stopwords as nltk_stopwords
from sklearn.metrics import f1_score
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from catboost import CatBoostClassifier


#Загрузим датасет
toxic_comments = pd.read_csv('/datasets/toxic_comments.csv')

#Отобразим первые 5 строк таблицы
display(toxic_comments.head(5))

#Посмотрим информацию о таблице
toxic_comments.info()

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 2 columns):
text     159571 non-null object
toxic    159571 non-null int64
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


### Лемматизация твитов

In [2]:
#Проведем лемматизацию твитов
m = Mystem() 

def lemmatize(text):
    lem = m.lemmatize(text)
    lem_join = "".join(lem)
    return lem_join
    
toxic_comments['lemm_text'] = toxic_comments['text'].apply(lemmatize)
toxic_comments = toxic_comments.drop(['text'], axis=1)
display(toxic_comments.head(5))

toxic                                          lemm_text
0      0  Explanation\nWhy the edits made under my usern...
1      0  D'aww! He matches this background colour I'm s...
2      0  Hey man, I'm really not trying to edit war. It...
3      0  "\nMore\nI can't make any real suggestions on ...
4      0  You, sir, are my hero. Any chance you remember...

### Определим признаки и целевой признак

In [3]:
target = toxic_comments['toxic']
features = toxic_comments.drop(['toxic'], axis=1)

features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.4, random_state=12345)
features_valid, features_test, target_valid, target_test = train_test_split(features_valid, target_valid, test_size=0.5,random_state=12345)

### Создадим мешок слов с учётом стоп-слов

In [4]:
#Создадим мешок слов с учётом стоп-слов
nltk.download('stopwords')
stopwords = set(nltk_stopwords.words('english'))
count_tf_idf = TfidfVectorizer(stop_words=stopwords) 

features_train = count_tf_idf.fit_transform(features_train['lemm_text'].values.astype('U'))
features_valid = count_tf_idf.transform(features_valid['lemm_text'].values.astype('U'))
features_test = count_tf_idf.transform(features_test['lemm_text'].values.astype('U'))
print(features_train.shape)
print(features_valid.shape)
print(features_test.shape)

[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


(95742, 139581)
(31914, 139581)
(31915, 139581)


### Проверим распределение между классами

In [5]:
#Посмотрим распределение между 0 и 1
display(toxic_comments['toxic'].value_counts())

0    143346
1     16225
Name: toxic, dtype: int64

In [6]:
#Так как распределение классов отличается более, чем в 8 раз. В этом случае изменим веса
class_ratio = toxic_comments['toxic'].value_counts()[0] / toxic_comments['toxic'].value_counts()[1]
class_weight={0:1, 1:class_ratio}

### Вывод:

Датасет был загружен, после чего произведена лемматизация твитов. Произведено определние признаков и целевого признака, псоле чего датасет разделен в пропорции 60\20\20. После проверки распределния между классами 0 и 1 выяснилось, что распределение отличается более чем в 8 раз. В сявзи с чем были изменены веса. 

## Обучение

### Линейная регрессия. Подбор гиперпараметров

In [7]:
#Подбор гиперпараметров
model_LR = LogisticRegression()
hparams = [{'solver':['newton-cg', 'lbfgs', 'liblinear'],
                'C':[0.1, 1, 10],
                'class_weight':[class_weight]}]

clf = GridSearchCV(model_LR, hparams, scoring='f1',cv=3)
clf.fit(features_train, target_train)
print("Лучшая модель LogisticRegression:\n")

LR_best_params = clf.best_params_
print(LR_best_params)
print()
print("Выборка гиперпараметров:\n")

means = clf.cv_results_['mean_test_score']
stds = clf.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, clf.cv_results_['params']):
    print("%0.6f for %r"% (mean, params))

cv_f1_LR = max(means)

/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)
/opt/conda

Лучшая модель LogisticRegression:

{'C': 10, 'class_weight': {0: 1, 1: 8.834884437596301}, 'solver': 'newton-cg'}

Выборка гиперпараметров:

0.711060 for {'C': 0.1, 'class_weight': {0: 1, 1: 8.834884437596301}, 'solver': 'newton-cg'}
0.711092 for {'C': 0.1, 'class_weight': {0: 1, 1: 8.834884437596301}, 'solver': 'lbfgs'}
0.711181 for {'C': 0.1, 'class_weight': {0: 1, 1: 8.834884437596301}, 'solver': 'liblinear'}
0.750942 for {'C': 1, 'class_weight': {0: 1, 1: 8.834884437596301}, 'solver': 'newton-cg'}
0.750823 for {'C': 1, 'class_weight': {0: 1, 1: 8.834884437596301}, 'solver': 'lbfgs'}
0.751038 for {'C': 1, 'class_weight': {0: 1, 1: 8.834884437596301}, 'solver': 'liblinear'}
0.764250 for {'C': 10, 'class_weight': {0: 1, 1: 8.834884437596301}, 'solver': 'newton-cg'}
0.763967 for {'C': 10, 'class_weight': {0: 1, 1: 8.834884437596301}, 'solver': 'lbfgs'}
0.764250 for {'C': 10, 'class_weight': {0: 1, 1: 8.834884437596301}, 'solver': 'liblinear'}


### Обучение модели LogisticRegression на подобранных гиперпараметрах

In [8]:
#Обучение модели LogisticRegression на подобранных гиперпараметрах:
model_LR = LogisticRegression()
model_LR.set_params(**LR_best_params)
model_LR.fit(features_train, target_train)
target_predict = model_LR.predict(features_valid)
valid_f1_LR = f1_score(target_valid, target_predict)
print('F1 модели LogisticRegression:', valid_f1_LR)

F1 модели LogisticRegression: 0.7707129094412332


### Модель DecisionTreeClassifier. Подбор гипперпараметров

In [9]:
model_DTC = DecisionTreeClassifier()
hparams = [{'max_depth':[x for x in range(20,50,3)],'random_state':[12345],'class_weight':[class_weight]}]

clf = GridSearchCV(model_DTC, hparams, scoring='f1',cv=3)
clf.fit(features_train, target_train)
print("Лучшие параметры модели DecisionTreeClassifier:\n")
DTC_best_params = clf.best_params_
print(DTC_best_params)

print("\nВыборка гиперпараметров:\n")
means = clf.cv_results_['mean_test_score']
stds = clf.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, clf.cv_results_['params']):
    print("%0.6f for %r"% (mean, params))
print()

cv_f1_DTC = max(means)

Лучшие параметры модели DecisionTreeClassifier:

{'class_weight': {0: 1, 1: 8.834884437596301}, 'max_depth': 47, 'random_state': 12345}

Выборка гиперпараметров:

0.596525 for {'class_weight': {0: 1, 1: 8.834884437596301}, 'max_depth': 20, 'random_state': 12345}
0.603322 for {'class_weight': {0: 1, 1: 8.834884437596301}, 'max_depth': 23, 'random_state': 12345}
0.611908 for {'class_weight': {0: 1, 1: 8.834884437596301}, 'max_depth': 26, 'random_state': 12345}
0.617481 for {'class_weight': {0: 1, 1: 8.834884437596301}, 'max_depth': 29, 'random_state': 12345}
0.613159 for {'class_weight': {0: 1, 1: 8.834884437596301}, 'max_depth': 32, 'random_state': 12345}
0.621410 for {'class_weight': {0: 1, 1: 8.834884437596301}, 'max_depth': 35, 'random_state': 12345}
0.613035 for {'class_weight': {0: 1, 1: 8.834884437596301}, 'max_depth': 38, 'random_state': 12345}
0.612439 for {'class_weight': {0: 1, 1: 8.834884437596301}, 'max_depth': 41, 'random_state': 12345}
0.619932 for {'class_weight': {0: 1, 

### Обучение модели DecisionTreeClassifier на подобранных гиперпараметрах

In [10]:
#Обучение модели DecisionTreeClassifier на подобранных гиперпараметрах:
model_DTC = DecisionTreeClassifier()
model_DTC.set_params(**DTC_best_params)
model_DTC.fit(features_train, target_train)
target_predict = model_DTC.predict(features_valid)
valid_f1_DTC = f1_score(target_valid, target_predict)
print('F1 модели DecisionTreeClassifier:', valid_f1_DTC)

F1 модели DecisionTreeClassifier: 0.6098719240178444


### Обучение модели CatBoostClassifier

In [11]:
#Модель CatBoostClassifier
model_CBC = CatBoostClassifier(verbose=False, iterations=230)
model_CBC.fit(features_train, target_train)
target_predict = model_CBC.predict(features_valid)
cv_f1_CBC = cross_val_score(model_CBC, features_train, target_train, cv=3, scoring='f1').mean()
valid_f1_CBC = f1_score(target_valid, target_predict)
print('F1 модели CatBoostClassifier:', valid_f1_CBC)

F1 модели CatBoostClassifier: 0.7427740410834394


### Вывод

Было обучено три модели: 

- Линейная регрессия
- Дерево решений
- CatBoostClassifier

Так как модель дерево решений имеет низкий показатель f1_score, а модель CatBoostClassifier долго обучается (около 1 часа) и имеет f1_score = 0.74277, что меньше 0.75, то для проверки моделей на тестовых данных будем использовать Линейную регрессию. 

## Выводы

### Проверка линейной регрессии на тестовых данных. 

In [11]:
model_LR = LogisticRegression()
model_LR.set_params(**LR_best_params)
model_LR.fit(features_train, target_train)
predict_test = model_LR.predict(features_test)
valid_f1_LR = f1_score(target_test, predict_test)
print('F1 модели LogisticRegression:', valid_f1_LR)

F1 модели LogisticRegression: 0.7697841726618705


### Вывод

Было обучено три модели: 

- Линейная регрессия
- Дерево решений
- CatBoostClassifier

Так как только Линейная регрессия имеет показатель f1_score больше 0.75, то эта модель была проверена на тестовых данных. В свзяи с тем, что TF-IDF превращает текст в числа, то лучшими моделью стала LogisticRegression, значения f1_score = 0.7697841726618705. Далее следует CatBoostClassifier, но он очень долго обучается (около часа) и на последнем месте Дерево решений. 

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны